File: src/Components/GitHub/GitHubApi.cs
namespace Components.GitHub
{
    using System;
    using System.Diagnostics.CodeAnalysis;
    using System.Net;
    using System.Net.Http;
    using System.Net.Http.Headers;
    using System.Text;
    using System.Threading.Tasks;
    using Components.GitHub.Dto;

    public class GitHubApi : IGitHubApi
    {
        private const string ApiBaseUri = @"https://api.github.com/";

        public async Task<string> GetSha(
            string userAgent,
            string authorizationToken,
            string repositoryName,
            string branchName)
        {
            HttpClient httpClient = new HttpClient { BaseAddress = new Uri(ApiBaseUri) };
            this.SetRequestHeaders(httpClient.DefaultRequestHeaders, userAgent, authorizationToken, null);

            var requestUri = string.Format(@"repos/{0}/{1}/git/refs/heads/{2}", userAgent, repositoryName, branchName);
            HttpResponseMessage response = await httpClient.GetAsync(requestUri).ConfigureAwait(false);
            response.EnsureSuccessStatusCode();

            var repo = await response.Content.ReadAsJsonAsync<RefResponse>().ConfigureAwait(false);
            return repo.Object.Sha;
        }

        public async Task CreateRepositoryBranch(
            string userAgent,
            string authorizationToken,
            string repositoryName,
            string masterBranchName,
            string newBranchName)
        {
            HttpClient httpClient = new HttpClient { BaseAddress = new Uri(ApiBaseUri) };
            this.SetRequestHeaders(httpClient.DefaultRequestHeaders, userAgent, authorizationToken, null);

            var requestUri = string.Format(@"repos/{0}/{1}/git/refs", userAgent, repositoryName);

            var sha = await this.GetSha(
                userAgent,
                authorizationToken,
                repositoryName,
                masterBranchName).ConfigureAwait(false);

            var branchRequest = new BranchRequest
            {
                Ref = string.Format(@"refs/heads/{0}", newBranchName),
                Sha = sha
            };

            HttpResponseMessage response = await httpClient.PostAsJsonAsync(requestUri, branchRequest).ConfigureAwait(false);
            response.EnsureSuccessStatusCode();
        }

        public async Task UpdateFile(
            string userAgent,
            string authorizationToken,
            string repositoryName,
            string branchName,
            string fileName,
            string content)
        {
            HttpClient httpClient = new HttpClient { BaseAddress = new Uri(ApiBaseUri) };
            this.SetRequestHeaders(httpClient.DefaultRequestHeaders, userAgent, authorizationToken, null);

            // get file to update
            var requestUri = string.Format(@"repos/{0}/{1}/contents/{2}?ref={3}", userAgent, repositoryName, fileName, branchName);
            HttpResponseMessage response = await httpClient.GetAsync(requestUri).ConfigureAwait(false);
            response.EnsureSuccessStatusCode();

            var fileToUpdate = await response.Content.ReadAsJsonAsync<FileContentResponse>().ConfigureAwait(false);
            byte[] data = Convert.FromBase64String(fileToUpdate.Content);
            string decodedData = Encoding.UTF8.GetString(data);

            // update file content
            decodedData += content;

            // update file
            requestUri = string.Format(@"repos/{0}/{1}/contents/{2}", userAgent, repositoryName, fileName);
            var newFile = new FileContentRequest
            {
                Message = "add comment",
                Content = Convert.ToBase64String(Encoding.UTF8.GetBytes(decodedData)),
                Sha = fileToUpdate.Sha,
                Branch = branchName
            };

            response = await httpClient.PutAsJsonAsync(requestUri, newFile).ConfigureAwait(false);
            response.EnsureSuccessStatusCode();
        }

        public async Task<string> CreatePullRequest(
            string userAgent,
            string authorizationToken,
            string repositoryName,
            string headBranchName,
            string baseBranchName)
        {
            HttpClient httpClient = new HttpClient { BaseAddress = new Uri(ApiBaseUri) };
            this.SetRequestHeaders(httpClient.DefaultRequestHeaders, userAgent, authorizationToken, null);

            var requestUri = string.Format(@"repos/{0}/{1}/pulls", userAgent, repositoryName);

            var pullRequest = new PullRequestRequest
            {
                Title = headBranchName,
                Body = "comment to merge",
                Head = headBranchName,
                Base = baseBranchName
            };

            HttpResponseMessage response = await httpClient.PostAsJsonAsync(requestUri, pullRequest).ConfigureAwait(false);
            response.EnsureSuccessStatusCode();

            return response.Headers.Location.AbsoluteUri;
        }

        [SuppressMessage("StyleCop.CSharp.SpacingRules", "SA1009:ClosingParenthesisMustBeSpacedCorrectly", Justification = "Reviewed.")]
        public async Task<(bool result, string etag)> IsPullRequestOpen(
            string userAgent,
            string authorizationToken,
            string pullRequestUrl,
            string etag)
        {
            HttpClient httpClient = new HttpClient { BaseAddress = new Uri(ApiBaseUri) };
            this.SetRequestHeaders(httpClient.DefaultRequestHeaders, userAgent, authorizationToken, etag);

            HttpResponseMessage response = await httpClient.GetAsync(pullRequestUrl).ConfigureAwait(false);

            if (response.StatusCode == HttpStatusCode.NotModified)
            {
                return (true, response.Headers.ETag.Tag);
            }

            if (response.StatusCode == HttpStatusCode.OK)
            {
                var pullRequestResponse = await response.Content.ReadAsJsonAsync<IsPullRequestExistsResponse>().ConfigureAwait(false);
                return (pullRequestResponse.State == @"open", response.Headers.ETag.Tag);
            }

            var exception = new HttpRequestException(string.Format(@"Response bad satus code: {0}", response.StatusCode));
            exception.Data.Add("response", response);
            throw exception;
        }

        [SuppressMessage("StyleCop.CSharp.SpacingRules", "SA1009:ClosingParenthesisMustBeSpacedCorrectly", Justification = "Reviewed.")]
        public async Task<bool> IsPullRequestMerged(
            string userAgent,
            string authorizationToken,
            string pullRequestUrl)
        {
            HttpClient httpClient = new HttpClient { BaseAddress = new Uri(ApiBaseUri) };

            ////ETag returned by github api is always null so it doesn't make sense to set it
            this.SetRequestHeaders(httpClient.DefaultRequestHeaders, userAgent, authorizationToken, null);

            var requestUri = string.Format(@"{0}/merge", pullRequestUrl);
            HttpResponseMessage response = await httpClient.GetAsync(requestUri).ConfigureAwait(false);

            if (response.StatusCode == HttpStatusCode.NoContent)
            {
                return true;
            }
            else if (response.StatusCode == HttpStatusCode.NotFound)
            {
                return false;
            }

            var exception = new HttpRequestException(string.Format(@"Response bad satus code: {0}", response.StatusCode));
            exception.Data.Add("response", response);
            throw exception;
        }

        private void SetRequestHeaders(HttpRequestHeaders httpRequestHeaders, string userAgent, string authorizationToken, string etag)
        {
            httpRequestHeaders.Accept.Clear();
            httpRequestHeaders.Add("User-agent", userAgent);
            httpRequestHeaders.Add("Authorization", string.Format("Token {0}", authorizationToken));

            if (!string.IsNullOrEmpty(etag))
            {
                httpRequestHeaders.Add("If-None-Match", string.Format(@"{0}", etag));
            }

            httpRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        }
    }
}

File: src/Components.Logic/GitHubApi.fs
module internal GitHubApi

open FSharp.Data
open System.Runtime.CompilerServices
open System
open System.Net
open System.Net.Http
open System.Text

[<assembly: InternalsVisibleTo("Components.Logic.FSharp.Integration.Tests")>]
do()

[<Literal>]
let ApiBaseUrl = @"https://api.github.com"

let getBaseHeaders userAgent authorizationToken (etag: string option) =
     let baseHeaders = ["User-agent", userAgent; "Authorization", "Token " + authorizationToken; "Accept", "application/json"]
     match etag with
     | Some etag -> baseHeaders |> List.append ["If-None-Match", etag]
     | None -> baseHeaders

module GetSha =
    [<Literal>]
    let Json = """
    {
	    "ref": "some_value",
	    "node_id": "some_value",
	    "url": "some_value",
	    "object":
        {
		    "sha": "some_value",
		    "type": "some_value",
		    "url": "some_value"
        }
    }
    """

    type Provider = JsonProvider<Json>

    let execute userAgent repositoryName branchName =
        async { 
            let url = sprintf @"%s/repos/%s/%s/git/refs/heads/%s" ApiBaseUrl userAgent repositoryName branchName
            let! response = url |> Provider.AsyncLoad
            return response.Object.Sha
        }

module GetFile =
    [<Literal>]
    let Json = """
    {
        "content": "some_value",
        "path": "some_value",
        "sha": "some_value"
    }
    """

    type Provider = JsonProvider<Json>

    let execute userAgent repositoryName fileName branchName =
        async { 
            let url = sprintf @"%s/repos/%s/%s/contents/%s?ref=%s" ApiBaseUrl userAgent repositoryName fileName branchName
            let! response = url |> Provider.AsyncLoad
            return response
        }

module CreateRepositoryBranch =
    [<Literal>]
    let Json = """
    {
	    "ref": "some_value",
	    "sha": "some_value"
    }
    """

    type Provider = JsonProvider<Json, RootName="branch">

    let execute userAgent authorizationToken repositoryName masterBranchName newBranchName =
        async {
            let url = sprintf @"%s/repos/%s/%s/git/refs" ApiBaseUrl userAgent repositoryName
            let headers = getBaseHeaders userAgent authorizationToken None
        
            let! sha = GetSha.execute userAgent repositoryName masterBranchName
            let post = Provider.Branch(@"refs/heads/" + newBranchName, sha)
            let! _ = post.JsonValue.RequestAsync (url = url, httpMethod = "POST", headers = headers)
            ()
        }

module UpdateFile = 
    [<Literal>]
    let Json = """
    {
        "message": "some_value",
        "content": "some_value",
        "sha": "some_value",
        "branch": "some_value"
    }
    """

    type Provider = JsonProvider<Json, RootName="newContent">
    
    let execute userAgent authorizationToken repositoryName fileName branchName content =
        async {
            let! file = GetFile.execute userAgent repositoryName fileName branchName
            let currentContent = file.Content |> Convert.FromBase64String |> Encoding.UTF8.GetString
            
            let newContent = currentContent + content
            let newContentBase64string = newContent |> Encoding.UTF8.GetBytes |> Convert.ToBase64String
            
            let url = sprintf @"%s/repos/%s/%s/contents/%s" ApiBaseUrl userAgent repositoryName fileName
            let headers = getBaseHeaders userAgent authorizationToken None

            let post = Provider.NewContent("add comment", newContentBase64string, file.Sha, branchName)
            let! _ = post.JsonValue.RequestAsync (url = url, httpMethod = "PUT", headers = headers)
            ()
        }

module CreatePullRequest =
    [<Literal>]
    let Json = """
    {
        "title": "some_value",
        "body": "some_value",
        "head": "some_value",
        "base": "some_value"
    }
    """

    type Provider = JsonProvider<Json, RootName="pull">

    let execute userAgent authorizationToken repositoryName headBranchName baseBranchName =
        async {
            let url = sprintf @"%s/repos/%s/%s/pulls" ApiBaseUrl userAgent repositoryName
            let headers = getBaseHeaders userAgent authorizationToken None

            let post = Provider.Pull(headBranchName, "comment to merge", headBranchName, baseBranchName)
            let! _ = post.JsonValue.RequestAsync (url = url, httpMethod = "POST", headers = headers)
            ()
        }

module IsPullRequestOpen =
    [<Literal>]
    let Json = """
    {
        "state": "some_value"
    }
    """

    type Provider = JsonProvider<Json>
    type result = {isOpen: bool; etag: string}

    let execute userAgent authorizationToken pullRequestUrl etag =
        async {
            let headers = getBaseHeaders userAgent authorizationToken etag
            let! response = Http.AsyncRequest(pullRequestUrl, headers = headers, silentHttpErrors = true)
            let etagHeader = response.Headers.TryFind "ETag"
            let etagValue =
                match etagHeader with
                | Some etag -> etag
                | None -> raise (ArgumentException("There is no ETag Header."))
            
            match response.StatusCode with
            | HttpStatusCodes.NotModified ->
                return {isOpen = true; etag = etagValue}
            | HttpStatusCodes.OK ->
                match response.Body with
                | Text body ->
                    let state = (Provider.Parse body).State
                    return {isOpen = (state = "open"); etag = etagValue}
                | _ -> return raise (ArgumentException("Invalid body format. Expected json as text."))
            | _ ->
                let ex = HttpRequestException(sprintf "Response bad status code: %d" response.StatusCode);
                ex.Data.Add("response", response);
                return raise ex
        }
        
module IsPullRequestMerged =
    let execute userAgent authorizationToken pullRequestUrl =
        async {
            let headers = getBaseHeaders userAgent authorizationToken None
            let url = sprintf "%s/merge" pullRequestUrl
            let! response = Http.AsyncRequest(url, headers = headers, silentHttpErrors = true)
            
            match response.StatusCode with
            | HttpStatusCodes.NoContent ->
                return true
            | HttpStatusCodes.NotFound ->
                return false
            | _ ->
                let ex = HttpRequestException(sprintf "Response bad status code: %d" response.StatusCode);
                ex.Data.Add("response", response);
                return raise ex
        }